vertx-mongo-client/src/main/java/io/vertx/ext/mongo/impl/MongoIterableStream.java

```diff
@@ -0,0 +1,186 @@
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * @author Thomas Segismont
+ */
+class MongoIterableStream implements ReadStream<JsonObject> {
+
+  private static final int BATCH_SIZE = 10;
+
+  private final Context context;
+  private final MongoIterable<JsonObject> mongoIterable;
+
+  private AsyncBatchCursor<JsonObject> batchCursor;
+  private Deque<JsonObject> queue;
+  private Handler<JsonObject> dataHandler;
+  private Handler<Throwable> exceptionHandler;
+  private Handler<Void> endHandler;
+  private boolean paused;
+  private boolean readInProgress;
+  private boolean closed;
+
+  MongoIterableStream(Context context, MongoIterable<JsonObject> mongoIterable) {
+    this.context = context;
+    this.mongoIterable = mongoIterable;
+  }
+
+  @Override
+  public synchronized MongoIterableStream exceptionHandler(Handler<Throwable> handler) {
+    checkClosed();
+    this.exceptionHandler = handler;
+    return this;
+  }
+
+  private void checkClosed() {
+    if (closed) {
+      throw new IllegalArgumentException("Stream is closed");
+    }
+  }
+
+  @Override
+  public synchronized MongoIterableStream handler(Handler<JsonObject> handler) {
+    checkClosed();
+    if (handler == null) {
+      close();
+    } else {
+      dataHandler = handler;
+      SingleResultCallback<AsyncBatchCursor<JsonObject>> callback = (result, t) -> {
+        context.runOnContext(v -> {
+          synchronized (this) {
+            if (t != null) {
+              close();
+              handleException(t);
+            } else {
+              batchCursor = result;
+              if (canRead()) {
+                doRead();
+              }
+            }
+          }
+        });
+      };
+      try {
+        mongoIterable.batchCursor(callback);
+      } catch (Exception e) {
+        close();
+        handleException(e);
+      }
+    }
+    return this;
+  }
+
+  private boolean canRead() {
+    return !paused && !closed;
+  }
+
+  @Override
+  public synchronized MongoIterableStream pause() {
+    checkClosed();
+    paused = true;
+    return this;
+  }
+
+  @Override
+  public synchronized MongoIterableStream resume() {
+    checkClosed();
+    if (paused) {
+      paused = false;
+      if (dataHandler != null) {
+        doRead();
+      }
+    }
+    return this;
+  }
+
+  private synchronized void doRead() {
+    if (readInProgress) {
+      return;
+    }
+    readInProgress = true;
+    if (queue == null) {
+      queue = new ArrayDeque<>(BATCH_SIZE);
+    }
+    if (!queue.isEmpty()) {
+      context.runOnContext(v -> emitQueued());
+      return;
+    }
+    batchCursor.setBatchSize(BATCH_SIZE);
+    context.<List<JsonObject>>executeBlocking(fut -> {
+      batchCursor.next((result, t) -> {
+        if (t != null) {
+          fut.fail(t);
+        } else {
+          fut.complete(result == null ? Collections.emptyList() : result);
+        }
+      });
+    }, false, ar -> {
+      synchronized (this) {
```

Comment:
You should lock on the right object.

Reply:
`this` references the stream instance

Reply:
comment this again for follow on developers or use an explicit lock mechanism